### ISSUE TITLE
Custom Notifications Viewer

### ISSUE THREAD

**Issue by 0x4007** [ubiquity-os/.github#1]
- I did some research in the past with custom notifications viewer. https://github.com/ubiquity/work.ubq.fi/pull/126
- Original specification here https://github.com/ubiquity/work.ubq.fi/issues/117

Fork work.ubq.fi because it has a lot of good qualities for list viewing, progressive web app features, styling etc. Tear out all the proposal/task related viewing logic, and focus only on notification viewing so that the code is simpler. 

## Custom Notifications Viewer

As activity continues to pick up and my notifications become slower to get through, it would be useful to make a simple ui that allows for more intelligent notification filtering. 

### Prioritization Sorting Algorithm

- We need to detect the priority level of the task the notification is associated with.
   - If the notification is not associated with a task, consider not showing it on this view at all. 
   - If the notification is a pull request, it must be linked to a task with a priority label.
- Reverse chronological order, meaning that neglected/old notifications should appear higher than new ones. 
   - The reason for this is that active conversations will continue to produce more notifications, which can easily cause the reviewer to check it multiple times in a sitting (as I try to get to inbox 0 this happens to me, so I intentionally start from older notifications for this reason.)
- Top of displayed list is highest priority, then work your way down to low priority. 

### Blocking Tasks

If it thinks without my input the task is blocked, it's a higher priority notification.
- Linkbacks: if the notification is associated with a task with many linkbacks then it should be considered higher priority, it usually means its blocking other pulls/tasks

Perhaps we could have an AI layer on top one day to summarize and/or reorder notifications from most important to least.

### Remarks
- Always exclude draft pulls.

## Notification Scoring

We should associate each notification with "scores" so that we can quantify where it would render on the page. 
- Time score: the older it is, the higher its score. We can take the current posix timestamp and subtract the timestamp of the notification to achieve this. 
- Linkback score: we can count how many times it was cross-referenced. The more times it was cross-referenced, the higher it should render on the page. We should prioritize those with higher linkbacks. If two notifications have the same amount of linkbacks i.e. 0, then we sort based on the time score only. 

We need to ensure that priority level always has precedence. For example, `priority 2` associated notifications should always appear higher than `priority 1`.

**Comment by 0x4007** [comment]
I decided to change this to 2 weeks because: 

Forking from work.ubq.fi and cleaning things up will take a day, then the actual start of the work will begin later in the week. I realize that the scoring systems: 1. timestamp 2. linkback and 3. priority level alone might take a week to get right. 

Timestamp is the least complex, priority linkback is probably next and then priority level is the most because it requires checking the associated issue and then reading its labels.

**Comment by zugdev** [comment]
btw I am implementing at https://github.com/ubiquity/notifications.ubq.fi

**Comment by zugdev** [comment]
these are all the reasons a notification can be received for:

![image](https://github.com/user-attachments/assets/96c48ae3-2a95-4618-8cf5-e9c8b54f05c9)

I thought of instantly filtering out: author, comment, ci_activity, invitation, member_feature_requested, security_advisory_credit, state_change, team_mention

**Comment by zugdev** [comment]
I am filtering out PR notifications which:

1. PR is in draft
2. PR is closed/merged
3. PR doesn't have a valid issue attached with Resolves

**Comment by 0x4007** [comment]
Seems like you're on the right track!

**Comment by 0x4007** [comment]
> btw I am implementing at https://github.com/ubiquity/notifications.ubq.fi

Normally you should be working in your own org, and in a pull, and link it here. Eventually you need to link a pull anyways for code related tasks

**Comment by zugdev** [comment]
good! just to walkthrough my thoughts:

we will need to have 2 different "elements":

1. PR notifications

for these we have to fetch the PR object and scrape the issue URL from "Resolves (url or #number)", then we fetch the issue object. I am then bundling: notification + pr + issue in an aggregated type. this will allow me to get  different properties from a single json object when rendering and caching: i.e issue labels, pr state, notification timestamp...

2. issue notifications

here we don't have to bundle PR, so it's a subset of the above.

---

features which could be cool in the future:

1. filtering out PR notifs where PR user is not the issue assignee

not yet implementable since disqualifier removes people

...

**Comment by 0x4007** [comment]
Actually I think we should cache all the backend details (if any are missing) in the GitHub.com/ubiquity/devpool-directory cron job

Right now in the __STORAGE__ branch I save all of the issue data which can be conveniently fetched in one shot. Don't handle these operations from the UI because it's a lot more efficient to cache and store in the repo.

**Comment by 0x4007** [comment]
@zugdev you need to link to prove progress to the OS

A note I wanted to drop here: it would be really nice to have a feature where we can essentially "dm" an issue of interest and have it displayed in a highlighted or highly visible section with the name of the sender next to it. 

A generic default UI we can model is an inbox with the persons GitHub profile and a timestamp. 

The thing is in practice I've noticed that the team tends to send issues to each other in telegram DMs which seems to work as a concise nudge. 

We can further clean our telegram dms by having this UI for this purpose.

**Comment by zugdev** [comment]
sick!

**Comment by zugdev** [comment]
It's a great idea! Will link the PR

**Comment by zugdev** [comment]
@0x4007 when you see this, pls tag me in the issue and PR so I can copy the objects for QA. my inbox is empty

**Comment by 0x4007** [comment]
@zugdev Issue tag here

**Comment by 0x4007** [comment]
Another later idea but we can make it a more formalized process for the necessary tags to be added. 

- Priority tags by project managers. 
- Time tags by technical team members. 

It would be interesting one day to have another "inbox" for tasks that are pending the other tag. 

For example, some have Time labels and then it's waiting on project managers to come in and set a Priority level so that it's funded. 

This in addition to a financial incentive (somewhere between $5 and $25 to set a label) can potentially move things along faster.

**Comment by zugdev** [comment]
The first one can and should be embedded into the notifications UI! I think it should later become a guide to our contributors, admins and managers. I see the final product as a efficiency and productivity maximizer, which funnels effort into the appropriate places. This way I believe it could become a product in itself, useful for partners, and not only an internal tool.

**Comment by 0x4007** [comment]
@gentlementlegen looks like the config is broken here

**Comment by gentlementlegen** [comment]
@0x4007 this is the only plugin listed in the configuration file: https://github.com/ubiquity-os/.github/blob/main/.github/.ubiquity-os.config.dev.yml

**Comment by 0x4007** [comment]
Right I see. The config merging is not implemented yet. Would you be able to render the full config for this repo?

**Comment by gentlementlegen** [comment]
Will have a look at the merging as soon as I am done with the current assigned issues I have. I will updated this to have the full config. We didn't deploy the configuration UI frotnend yet?

**Comment by 0x4007** [comment]
> Will have a look at the merging as soon as I am done with the current assigned issues I have. I will updated this to have the full config. We didn't deploy the configuration UI frotnend yet?

I don't think it's mapped to a URL yet, just some cloudflare link

**Comment by 0x4007** [comment]
@zugdev The login redirects to Devpool directory please fix

**Comment by zugdev** [comment]
> @zugdev The login redirects to Devpool directory please fix

I will add the regex to supabase if that's fine.

**Comment by 0x4007** [comment]
Can you replicate the rounded squares we have at directory for the avatars 

![image](https://github.com/user-attachments/assets/1ddad36b-70d7-476a-b6e6-ba1df6d6c8b3)

**Comment by zugdev** [comment]
@gentlementlegen Looks like the permit URLs weren't properly generated? They are undefined.

**Comment by 0x4007** [comment]
We might have to roll back to an earlier commit before we were rushing to fix for the demo.

**Comment by gentlementlegen** [comment]
https://github.com/ubiquity-os-marketplace/text-conversation-rewards/actions/runs/13336736054/job/37253537335#step:3:4457

The key was not allowed to generate permits here, I do not know what was encoded in that key.

**Comment by 0x4007** [comment]
I didn't change the key in our secrets or anywhere.

**Comment by gentlementlegen** [comment]
Didnt you change the evm encrypted key at some point?

**Comment by 0x4007** [comment]
Not for the demo

**Comment by zugdev** [comment]
Can we regenerate this? @gentlementlegen

**Comment by gentlementlegen** [comment]
@zugdev I will try to regen but if the key does not allow for generating here we will need to update the key itself 😄

**Comment by gentlementlegen** [comment]
The problem is still that the private key does not allow for permit generations in this repository it seems: https://github.com/ubiquity-os-marketplace/text-conversation-rewards/actions/runs/13601140899/job/38027114423#step:3:4604

**Comment by 0x4007** [comment]
The permits shouldn't render at all if they are invalid. It should throw an error like a warning and explain the problem with the key.

**Comment by gentlementlegen** [comment]
Just the URL shouldn't show, the permits are only one module that can be disabled so no error should be thrown. Maybe we could add a footnote with the error however.

**Comment by zugdev** [comment]
hey, this should be regenerated sometime

**Comment by 0x4007** [comment]
@gentlementlegen permit is broken?

**Comment by 0x4007** [comment]
2 weeks isn't standard I realized it shouldn't have existed.

**Comment by 0x4007** [comment]
I will need to temporarily add 2 weeks back into the config for this.
